designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+
+2. Make the stack allocator conform to accept `vari_type<T>` objects.
+
+3. Additional PRs for the current arithmetic operators for `var` to accept `vari_type<T>` object types.
+
+4. Sort out and add any additional specialized methods for static matrix types such as `cholesky_decompose`.
+
+5. Add the `static_matrix` and `static_vector` types to the Stan language along with supported signatures.
+
+6. Support detection and substitution of `matrix` types for `static_matrix` types.
+
+7. ???
+
+8. Done!
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+More templates can be confusing and will lead to larger compile time. The confusion of templates can be mitigated by proper duty of care with documentation and guides for the Stan Math type system. Larger compile times can't be avoided with this implementation, though other forthcoming proposals can allow us to monitor increases in compilation times.
+
+With `sparse_matrix` and `complex_matrix` this will now add _another_ `*_matrix` type proposal to the language. There's no way to get around this in the Stan language currently, though some small side discussions exist to support attributes on types such as
+
+```stan
+@(sparse_type, static_type) // can't have static here because of C++ keyword
+matrix[N, M] X;
+
+@(complex_type, static_type)
+vector[N] Y;
+```
+
+though this is far down the line for now.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Besides the tech burden I'm rather happy with this implementation, though I'm open to any criticisms/alternatives and will add them here. You can see some of the discussion on this as well in issue [#1805](https://github.com/stan-dev/math/issues/1805) and the prior art below.
+
+# Prior art
+[prior-art]: #prior-art
+
+Discussions:
+ - [Efficient `static_matrix` type?](https://discourse.mc-stan.org/t/efficient-static-matrix-type/2136)
+ - [Static/immutable matrices w. comprehensions](https://discourse.mc-stan.org/t/static-immutable-matrices-w-comprehensions/12641)
+ - [Stan SIMD and Performance](https://discourse.mc-stan.org/t/stan-simd-performance/10488/11)
+ - [A New Continuation Based Autodiff By Refactoring](https://discourse.mc-stan.org/t/a-new-continuation-based-autodiff-by-refactoring/5037/2)
+
+[Enoki](https://github.com/mitsuba-renderer/enoki) is a very nice C++17 library for automatic differentiation which under the hood can transform their autodiff type from an arrays of structs to structs of arrays. It's pretty neat! Though implementing something like their methods would require some very large changes to the way we handle automatic differentiation.
+
```

Comment:
JAX is the major system that's built like this, though I think PyTorch may also use a static design.  As far as I know, they *only* have static matrices.

Thanks for the pointer to Enoki.  I don't know that one (though I know the mushroom).

Reply:
Good call I'll link to these

Reply:
@bob-carpenter can you point me to any docs from jax / pytorch that talk about the sort of static matrix type they have? It looks like in pytorch they mainly use tensors and allow for subsetting

https://pytorch.org/docs/stable/tensor_view.html

Reply:
For JAX:

https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#🔪-In-Place-Updates

I love that they literally put a knife in the URL for the "sharp bits" of the doc.

There's no reason not to *use* a slice, block, or element of a matrix as an rvalue---the problems start when you assign to a matrix element.  That's why it's OK to assign whole matrices.  

Not allowing assignment of whole matrices means that transformed data, transformed parameters, and generated quantities need to be declare-defined in one line, which means they can't depend on any other calculations in the blocks they are declared in.  For example, the following code pattern becomes impossible.

```
transformed parameters {
  static_matrix a[3, 2] x; 
  T1 b;
  T2 c;
  ... some calculations involving b and c ...
  x = ... some function of those calculations
}
```

Instead, b and c need to be declare-defined before x and a function defined to call it, 

```
  matrix b = foo();
  matrix c = bar(b);
  matrix x = f(b, c);
```

and often it means we can't share computations that we'd like to share for calculating c or x.  At least that's been the problem in the past with defining sizes, which have this constant character.

Reply:
Added JAX doc to the prior art!

We are pushing the language discussion out of this design for now but seeing this I'm get this now